ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java

```diff
@@ -312,16 +321,16 @@ short getVersion() {
    * @return The blob index value that represents the key or null if not found
    * @throws StoreException
    */
-  IndexValue find(StoreKey keyToFind) throws StoreException {
-    IndexValue toReturn = null;
+  NavigableSet<IndexValue> find(StoreKey keyToFind) throws StoreException {
+    NavigableSet<IndexValue> toReturn = null;
     try {
       rwLock.readLock().lock();
       if (!mapped.get()) {
-        IndexValue value = index.get(keyToFind);
-        if (value != null) {
+        NavigableSet<IndexValue> values = index.get(keyToFind);
```

Comment:
This method should not return a direct reference to the associated set as the set may be modified concurrently while being iterated over by the caller. The problem is not with the concurrent access, but with the API semantics - callers would want a snapshot, phantom reads would be unexpected.

You could use `ConcurrentSkipListSet::clone()` method that does a shallow copy:
```java
if (values != null) {
  ...
  toReturn = values.clone();
} else {
  toReturn = null;
}
```

Reply:
Excellent point - will make the change

Reply:
Done

Reply:
Just want to make sure my understanding is correct:

Is it because new entries can be added when the set are being iterated?
`clone()` is used because we don't need to worried about new added entires? What if a delete entry is added for the same key? Because we only guarantee eventually consistency?